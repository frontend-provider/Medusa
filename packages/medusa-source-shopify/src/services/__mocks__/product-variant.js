export const ProductVariantServiceMock = {
  withTransaction: function () {
    return this
  },
  create: jest.fn().mockImplementation((_data) => {
    return Promise.resolve({})
  }),
  update: jest.fn().mockImplementation((_data) => {
    return Promise.resolve()
  }),
  delete: jest.fn().mockImplementation((_data) => {
    return Promise.resolve()
  }),
  addOption: jest.fn().mockImplementation((id, title) => {
    return Promise.resolve()
  }),
}
