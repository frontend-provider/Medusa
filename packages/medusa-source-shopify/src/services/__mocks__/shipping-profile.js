export const ShippingProfileServiceMock = {
  retrieveGiftCardDefault: jest.fn().mockImplementation((_data) => {
    return Promise.resolve({ id: "gift_card_profile" })
  }),
  retrieveDefault: jest.fn().mockImplementation((_data) => {
    return Promise.resolve({ id: "default_shipping_profile" })
  }),
}
