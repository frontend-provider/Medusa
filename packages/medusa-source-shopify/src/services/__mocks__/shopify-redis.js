export const ShopifyRedisServiceMock = {
  addIgnore: jest.fn().mockImplementation((_id, _event) => {
    return Promise.resolve()
  }),
  shouldIgnore: jest.fn().mockImplementation((_id, _event) => {
    return false
  }),
  addUniqueValue: jest.fn().mockImplementation((val, type) => {
    return Promise.resolve()
  }),
  getUniqueValue: jest.fn().mockImplementation((val, type) => {
    if (val === "uniq") {
      return Promise.resolve("uniq")
    } else {
      return Promise.resolve(undefined)
    }
  }),
}
