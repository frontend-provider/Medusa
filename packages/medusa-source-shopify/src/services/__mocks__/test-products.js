export const medusaProducts = {
  ipod: {
    id: "prod_ipod",
    title: "IPod Nano - 8GB",
    subtitle: null,
    description:
      "<p>It's the small iPod with one very big idea: Video. Now the world's most popular music player, available in 4GB and 8GB models, lets you enjoy TV shows, movies, video podcasts, and more. The larger, brighter display means amazing picture quality. In six eye-catching colors, iPod nano is stunning all around. And with models starting at just $149, little speaks volumes.</p>",
    handle: "ipod-nano",
    is_giftcard: false,
    status: "proposed",
    images: [
      {
        id: "img_01FHZ9ZFC4SP5ME8T18SW47R7S",
        url:
          "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano.png?v=1633120966",
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: null,
      },
      {
        id: "img_01FHZ9ZFC4SK54GE072FHMQ5XT",
        url:
          "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano-2.png?v=1633120966",
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: null,
      },
      {
        id: "img_01FHZ9ZFC44EKQPVWK76HBZT0S",
        url:
          "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano.png?v=1633120966",
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: null,
      },
    ],
    thumbnail:
      "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano.png?v=1633120966",
    options: [
      {
        id: "opt_01FHZ9ZFCPCQ7B1MPDD9X9YQX3",
        title: "Color",
        product_id: "prod_01FHZ9ZFC3KKYKA35NXNJ5A7FR",
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: null,
      },
    ],
    variants: [
      {
        id: "variant_01FHZ9ZFED5MEX3V4A7H5V2A0C",
        title: "Pink",
        product_id: "prod_01FHZ9ZFC3KKYKA35NXNJ5A7FR",
        prices: [
          {
            id: "ma_01FHZ9ZFEPN276FGACR08ZBZ2M",
            currency_code: "usd",
            amount: 19900,
            sale_amount: null,
            variant_id: "variant_01FHZ9ZFED5MEX3V4A7H5V2A0C",
            region_id: null,
            created_at: "2021-10-14T11:46:10.391Z",
            updated_at: "2021-10-14T11:46:10.391Z",
            deleted_at: null,
          },
        ],
        sku: "IPOD2008PINK",
        barcode: "1234_pink",
        ean: null,
        upc: "1234_pink",
        inventory_quantity: 10,
        allow_backorder: true,
        manage_inventory: true,
        hs_code: null,
        origin_country: null,
        mid_code: null,
        material: null,
        weight: 567,
        length: null,
        height: null,
        width: null,
        options: [
          {
            id: "optval_01FHZ9ZFEDMCJME1BR4X8H89MG",
            value: "Pink",
            option_id: "opt_01FHZ9ZFCPCQ7B1MPDD9X9YQX3",
            variant_id: "variant_01FHZ9ZFED5MEX3V4A7H5V2A0C",
            created_at: "2021-10-14T11:46:10.391Z",
            updated_at: "2021-10-14T11:46:10.391Z",
            deleted_at: null,
            metadata: null,
          },
        ],
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: {
          sh_id: 808950810,
        },
      },
      {
        id: "variant_01FHZ9ZFFA2KPA4JJSKF07VGWH",
        title: "Red",
        product_id: "prod_01FHZ9ZFC3KKYKA35NXNJ5A7FR",
        prices: [
          {
            id: "ma_01FHZ9ZFFFX31Z9GZ8ZHQ5KQ2P",
            currency_code: "usd",
            amount: 19900,
            sale_amount: null,
            variant_id: "variant_01FHZ9ZFFA2KPA4JJSKF07VGWH",
            region_id: null,
            created_at: "2021-10-14T11:46:10.391Z",
            updated_at: "2021-10-14T11:46:10.391Z",
            deleted_at: null,
          },
        ],
        sku: "IPOD2008RED",
        barcode: "1234_red",
        ean: null,
        upc: "1234_red",
        inventory_quantity: 20,
        allow_backorder: true,
        manage_inventory: true,
        hs_code: null,
        origin_country: null,
        mid_code: null,
        material: null,
        weight: 567,
        length: null,
        height: null,
        width: null,
        options: [
          {
            id: "optval_01FHZ9ZFFASNAD8JSWS85Y4ZM0",
            value: "Red",
            option_id: "opt_01FHZ9ZFCPCQ7B1MPDD9X9YQX3",
            variant_id: "variant_01FHZ9ZFFA2KPA4JJSKF07VGWH",
            created_at: "2021-10-14T11:46:10.391Z",
            updated_at: "2021-10-14T11:46:10.391Z",
            deleted_at: null,
            metadata: null,
          },
        ],
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: {
          sh_id: 49148385,
        },
      },
      {
        id: "variant_01FHZ9ZFG2F54TZMAWV0RJCQ5D",
        title: "Green",
        product_id: "prod_01FHZ9ZFC3KKYKA35NXNJ5A7FR",
        prices: [
          {
            id: "ma_01FHZ9ZFG6QPPCG4XPX8JG0HGW",
            currency_code: "usd",
            amount: 19900,
            sale_amount: null,
            variant_id: "variant_01FHZ9ZFG2F54TZMAWV0RJCQ5D",
            region_id: null,
            created_at: "2021-10-14T11:46:10.391Z",
            updated_at: "2021-10-14T11:46:10.391Z",
            deleted_at: null,
          },
        ],
        sku: "IPOD2008GREEN",
        barcode: "1234_green",
        ean: null,
        upc: "1234_green",
        inventory_quantity: 30,
        allow_backorder: true,
        manage_inventory: true,
        hs_code: null,
        origin_country: null,
        mid_code: null,
        material: null,
        weight: 567,
        length: null,
        height: null,
        width: null,
        options: [
          {
            id: "optval_01FHZ9ZFG2K7GG7B79VVZW6AKN",
            value: "Green",
            option_id: "opt_01FHZ9ZFCPCQ7B1MPDD9X9YQX3",
            variant_id: "variant_01FHZ9ZFG2F54TZMAWV0RJCQ5D",
            created_at: "2021-10-14T11:46:10.391Z",
            updated_at: "2021-10-14T11:46:10.391Z",
            deleted_at: null,
            metadata: null,
          },
        ],
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: {
          sh_id: 39072856,
        },
      },
      {
        id: "variant_01FHZ9ZFGN4GZN3P6JEB39C9G8",
        title: "Black",
        product_id: "prod_01FHZ9ZFC3KKYKA35NXNJ5A7FR",
        prices: [
          {
            id: "ma_01FHZ9ZFGVNSHF4YRANBJSVN11",
            currency_code: "usd",
            amount: 19900,
            sale_amount: null,
            variant_id: "variant_01FHZ9ZFGN4GZN3P6JEB39C9G8",
            region_id: null,
            created_at: "2021-10-14T11:46:10.391Z",
            updated_at: "2021-10-14T11:46:10.391Z",
            deleted_at: null,
          },
        ],
        sku: "IPOD2008BLACK",
        barcode: "1234_black",
        ean: null,
        upc: "1234_black",
        inventory_quantity: 40,
        allow_backorder: true,
        manage_inventory: true,
        hs_code: null,
        origin_country: null,
        mid_code: null,
        material: null,
        weight: 567,
        length: null,
        height: null,
        width: null,
        options: [
          {
            id: "optval_01FHZ9ZFGNJXHVHFERQHR1AXHX",
            value: "Black",
            option_id: "opt_01FHZ9ZFCPCQ7B1MPDD9X9YQX3",
            variant_id: "variant_01FHZ9ZFGN4GZN3P6JEB39C9G8",
            created_at: "2021-10-14T11:46:10.391Z",
            updated_at: "2021-10-14T11:46:10.391Z",
            deleted_at: null,
            metadata: null,
          },
        ],
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: {
          sh_id: 457924702,
        },
      },
    ],
    profile_id: "sp_01FHZ9YRZT5DD4HTKJG2T38WXD",
    weight: null,
    length: null,
    height: null,
    width: null,
    hs_code: null,
    origin_country: null,
    mid_code: null,
    material: null,
    collection_id: null,
    collection: null,
    type_id: null,
    type: null,
    tags: [
      {
        id: "ptag_01FHZ9ZFBVN7EK9VZ2EC0T2Y1A",
        value: "Emotive",
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: null,
      },
      {
        id: "ptag_01FHZ9ZFBY01226XFYXP0R3EEK",
        value: " Flash Memory",
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: null,
      },
      {
        id: "ptag_01FHZ9ZFC0N8S7G85F2TZXKSAM",
        value: " MP3",
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: null,
      },
      {
        id: "ptag_01FHZ9ZFC2K6V0A43S398DC2Q2",
        value: " Music",
        created_at: "2021-10-14T11:46:10.391Z",
        updated_at: "2021-10-14T11:46:10.391Z",
        deleted_at: null,
        metadata: null,
      },
    ],
    discountable: true,
    external_id: "shopify_ipod",
    created_at: "2021-10-14T11:46:10.391Z",
    updated_at: "2021-10-14T11:46:10.391Z",
    deleted_at: null,
    metadata: null,
  },
}

const shopifyIpod = {
  id: "shopify_ipod",
  title: "IPod Nano - 8GB",
  body_html:
    "<p>It's the small iPod with one very big idea: Video. Now the world's most popular music player, available in 4GB and 8GB models, lets you enjoy TV shows, movies, video podcasts, and more. The larger, brighter display means amazing picture quality. In six eye-catching colors, iPod nano is stunning all around. And with models starting at just $149, little speaks volumes.</p>",
  vendor: "Apple",
  product_type: "Cult Products",
  created_at: "2021-10-01T16:42:46-04:00",
  handle: "ipod-nano",
  updated_at: "2021-10-01T16:42:46-04:00",
  published_at: "2007-12-31T19:00:00-05:00",
  template_suffix: null,
  status: "active",
  published_scope: "web",
  tags: "Emotive, Flash Memory, MP3, Music",
  admin_graphql_api_id: "gid://shopify/Product/632910392",
  variants: [
    {
      id: 808950810,
      product_id: 632910392,
      title: "Pink",
      price: "199.00",
      sku: "IPOD2008PINK",
      position: 1,
      inventory_policy: "continue",
      compare_at_price: null,
      fulfillment_service: "manual",
      inventory_management: "shopify",
      option1: "Pink",
      option2: null,
      option3: null,
      created_at: "2021-10-01T16:42:46-04:00",
      updated_at: "2021-10-01T16:42:46-04:00",
      taxable: true,
      barcode: "1234_pink",
      grams: 567,
      image_id: 562641783,
      weight: 1.25,
      weight_unit: "lb",
      inventory_item_id: 808950810,
      inventory_quantity: 10,
      old_inventory_quantity: 10,
      presentment_prices: [
        {
          price: {
            amount: "199.00",
            currency_code: "USD",
          },
          compare_at_price: null,
        },
      ],
      requires_shipping: true,
      admin_graphql_api_id: "gid://shopify/ProductVariant/808950810",
    },
    {
      id: 49148385,
      product_id: 632910392,
      title: "Red",
      price: "199.00",
      sku: "IPOD2008RED",
      position: 2,
      inventory_policy: "continue",
      compare_at_price: null,
      fulfillment_service: "manual",
      inventory_management: "shopify",
      option1: "Red",
      option2: null,
      option3: null,
      created_at: "2021-10-01T16:42:46-04:00",
      updated_at: "2021-10-01T16:42:46-04:00",
      taxable: true,
      barcode: "1234_red",
      grams: 567,
      image_id: null,
      weight: 1.25,
      weight_unit: "lb",
      inventory_item_id: 49148385,
      inventory_quantity: 20,
      old_inventory_quantity: 20,
      presentment_prices: [
        {
          price: {
            amount: "199.00",
            currency_code: "USD",
          },
          compare_at_price: null,
        },
      ],
      requires_shipping: true,
      admin_graphql_api_id: "gid://shopify/ProductVariant/49148385",
    },
    {
      id: 39072856,
      product_id: 632910392,
      title: "Green",
      price: "199.00",
      sku: "IPOD2008GREEN",
      position: 3,
      inventory_policy: "continue",
      compare_at_price: null,
      fulfillment_service: "manual",
      inventory_management: "shopify",
      option1: "Green",
      option2: null,
      option3: null,
      created_at: "2021-10-01T16:42:46-04:00",
      updated_at: "2021-10-01T16:42:46-04:00",
      taxable: true,
      barcode: "1234_green",
      grams: 567,
      image_id: null,
      weight: 1.25,
      weight_unit: "lb",
      inventory_item_id: 39072856,
      inventory_quantity: 30,
      old_inventory_quantity: 30,
      presentment_prices: [
        {
          price: {
            amount: "199.00",
            currency_code: "USD",
          },
          compare_at_price: null,
        },
      ],
      requires_shipping: true,
      admin_graphql_api_id: "gid://shopify/ProductVariant/39072856",
    },
    {
      id: 457924702,
      product_id: 632910392,
      title: "Black",
      price: "199.00",
      sku: "IPOD2008BLACK",
      position: 4,
      inventory_policy: "continue",
      compare_at_price: null,
      fulfillment_service: "manual",
      inventory_management: "shopify",
      option1: "Black",
      option2: null,
      option3: null,
      created_at: "2021-10-01T16:42:46-04:00",
      updated_at: "2021-10-01T16:42:46-04:00",
      taxable: true,
      barcode: "1234_black",
      grams: 567,
      image_id: null,
      weight: 1.25,
      weight_unit: "lb",
      inventory_item_id: 457924702,
      inventory_quantity: 40,
      old_inventory_quantity: 40,
      presentment_prices: [
        {
          price: {
            amount: "199.00",
            currency_code: "USD",
          },
          compare_at_price: null,
        },
      ],
      requires_shipping: true,
      admin_graphql_api_id: "gid://shopify/ProductVariant/457924702",
    },
  ],
  options: [
    {
      id: 594680422,
      product_id: 632910392,
      name: "Color",
      position: 1,
      values: ["Pink", "Red", "Green", "Black"],
    },
  ],
  images: [
    {
      id: 850703190,
      product_id: 632910392,
      position: 1,
      created_at: "2021-10-01T16:42:46-04:00",
      updated_at: "2021-10-01T16:42:46-04:00",
      alt: null,
      width: 123,
      height: 456,
      src:
        "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano.png?v=1633120966",
      variant_ids: [],
      admin_graphql_api_id: "gid://shopify/ProductImage/850703190",
    },
    {
      id: 562641783,
      product_id: 632910392,
      position: 2,
      created_at: "2021-10-01T16:42:46-04:00",
      updated_at: "2021-10-01T16:42:46-04:00",
      alt: null,
      width: 123,
      height: 456,
      src:
        "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano-2.png?v=1633120966",
      variant_ids: [808950810],
      admin_graphql_api_id: "gid://shopify/ProductImage/562641783",
    },
    {
      id: 378407906,
      product_id: 632910392,
      position: 3,
      created_at: "2021-10-01T16:42:46-04:00",
      updated_at: "2021-10-01T16:42:46-04:00",
      alt: null,
      width: 123,
      height: 456,
      src:
        "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano.png?v=1633120966",
      variant_ids: [],
      admin_graphql_api_id: "gid://shopify/ProductImage/378407906",
    },
  ],
  image: {
    id: 850703190,
    product_id: 632910392,
    position: 1,
    created_at: "2021-10-01T16:42:46-04:00",
    updated_at: "2021-10-01T16:42:46-04:00",
    alt: null,
    width: 123,
    height: 456,
    src:
      "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano.png?v=1633120966",
    variant_ids: [],
    admin_graphql_api_id: "gid://shopify/ProductImage/850703190",
  },
}

export const shopifyProducts = {
  ipod: shopifyIpod,
  new_ipod: { ...shopifyIpod, id: "shopify_new_ipod" },
  ipod_update: {
    ...shopifyIpod,
    title: "IPod Nano",
    options: [
      ...shopifyIpod.options,
      {
        id: 594680423,
        product_id: 632910392,
        name: "Memory",
        position: 2,
        values: ["8GB", "16GB"],
      },
    ],
    variants: [
      { ...shopifyIpod.variants[0], title: "Pink / 8GB", option2: "8GB" },
      { ...shopifyIpod.variants[1], title: "Red / 8GB", option2: "8GB" },
      { ...shopifyIpod.variants[2], title: "Green / 8GB", option2: "8GB" },
      { ...shopifyIpod.variants[3], title: "Black / 8GB", option2: "8GB" },
      { ...shopifyIpod.variants[0], title: "Pink / 16GB", option2: "16GB" },
      { ...shopifyIpod.variants[1], title: "Red / 16GB", option2: "16GB" },
      { ...shopifyIpod.variants[2], title: "Green / 16GB", option2: "16GB" },
      { ...shopifyIpod.variants[3], title: "Black / 16GB", option2: "16GB" },
    ],
  },
}
