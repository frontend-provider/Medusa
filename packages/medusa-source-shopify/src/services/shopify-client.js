import { DataType } from "@shopify/shopify-api"
import { BaseService } from "medusa-interfaces"
import { createClient } from "../utils/create-client"
import { pager } from "../utils/pager"

class ShopifyClientService extends BaseService {
  // eslint-disable-next-line no-empty-pattern
  constructor({}, options) {
    super()

    this.options = options

    /** @private @const {ShopifyRestClient} */
    this.client_ = createClient(this.options)
  }

  get(params) {
    return this.client_.get(params)
  }

  async list(path, extraHeaders = null, extraQuery = {}) {
    return await pager(this.client_, path, extraHeaders, extraQuery)
  }

  delete(params) {
    return this.client_.delete(params)
  }

  post(params) {
    return this.client_.post({
      path: params.path,
      body: params.body,
      type: DataType.JSON,
    })
  }

  put(params) {
    return this.client_.post(params)
  }
}

export default ShopifyClientService
