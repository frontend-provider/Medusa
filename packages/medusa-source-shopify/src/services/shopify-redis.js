// shopify-redis
import { BaseService } from "medusa-interfaces"
import { IGNORE_THRESHOLD } from "../utils/const"

class shopifyRedisService extends BaseService {
  constructor({ redisClient }, options) {
    super()

    this.options_ = options

    /** @private @const {RedisClient} */
    this.redis_ = redisClient
  }

  async addIgnore(id, side) {
    const key = `sh_${id}_ignore_${side}`
    return await this.redis_.set(
      key,
      1,
      "EX",
      this.options_.ignore_threshold || IGNORE_THRESHOLD
    )
  }

  async shouldIgnore(id, action) {
    const key = `sh_${id}_ignore_${action}`
    return await this.redis_.get(key)
  }

  async addUniqueValue(uniqueVal, type) {
    const key = `sh_${uniqueVal}_${type}`
    return await this.redis_.set(key, 1, "EX", 60 * 5)
  }

  async getUniqueValue(uniqueVal, type) {
    const key = `sh_${uniqueVal}_${type}`
    return await this.redis_.get(key)
  }
}

export default shopifyRedisService
