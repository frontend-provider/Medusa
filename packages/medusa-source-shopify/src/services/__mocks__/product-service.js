import { medusaProducts } from "./test-products"

export const ProductServiceMock = {
  withTransaction: function () {
    return this
  },
  create: jest.fn().mockImplementation((data) => {
    if (data.handle === "ipod-nano") {
      return Promise.resolve(medusaProducts.ipod)
    }
  }),
  update: jest.fn().mockImplementation((_id, _update) => {
    return Promise.resolve(medusaProducts.ipod)
  }),
  retrieveByExternalId: jest.fn().mockImplementation((id) => {
    if (id === "shopify_ipod") {
      return Promise.resolve(medusaProducts.ipod)
    }
    if (id === "shopify_deleted") {
      return Promise.resolve(medusaProducts.ipod)
    }
    return Promise.resolve(undefined)
  }),
  retrieve: jest.fn().mockImplementation((_id, _config) => {
    if (_id === "prod_ipod") {
      return Promise.resolve(medusaProducts.ipod)
    }
  }),
  addOption: jest.fn().mockImplementation((_id, _title) => {
    return Promise.resolve()
  }),
}
