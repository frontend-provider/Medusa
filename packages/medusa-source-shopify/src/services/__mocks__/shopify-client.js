import { shopifyProducts } from "./test-products"

export const ShopifyClientServiceMock = {
  get: jest.fn().mockImplementation((params) => {
    if (params.path === "products/ipod") {
      return Promise.resolve(shopifyProducts.ipod)
    }
    if (params.path === "products/new_ipod") {
      return Promise.resolve(shopifyProducts.new_ipod)
    }
    if (params.path === "products/shopify_ipod") {
      return Promise.resolve({
        body: {
          product: shopifyProducts.ipod_update,
        },
      })
    }
    if (params.path === "products/shopify_deleted") {
      return Promise.resolve({
        body: {
          product: {
            ...shopifyProducts.ipod,
            variants: shopifyProducts.ipod.variants.slice(1, -1),
          },
        },
      })
    }
  }),
  list: jest.fn().mockImplementation((path, _headers, _query) => {
    if (path === "products") {
      return Promise.resolve([shopifyProducts.ipod])
    }
  }),
}
