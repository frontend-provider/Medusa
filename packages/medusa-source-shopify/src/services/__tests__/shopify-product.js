import { MockManager } from "medusa-test-utils"
import ShopifyProductService from "../shopify-product"
import { ProductServiceMock } from "../__mocks__/product-service"
import { ProductVariantServiceMock } from "../__mocks__/product-variant"
import { ShippingProfileServiceMock } from "../__mocks__/shipping-profile"
import { ShopifyClientServiceMock } from "../__mocks__/shopify-client"
import { ShopifyRedisServiceMock } from "../__mocks__/shopify-redis"
import { medusaProducts, shopifyProducts } from "../__mocks__/test-products"

describe("ShopifyProductService", () => {
  describe("normalizeProduct_", () => {
    const shopifyProductService = new ShopifyProductService({
      manager: MockManager,
      shopifyClientService: ShopifyClientServiceMock,
    })

    beforeEach(async () => {
      jest.clearAllMocks()
    })

    it("succesfully normalizes a product from Shopify", async () => {
      const data = await ShopifyClientServiceMock.get({ path: "products/ipod" })

      const normalized = shopifyProductService.normalizeProduct_(data)

      expect(normalized).toMatchSnapshot()
    })
  })

  describe("create", () => {
    const shopifyProductService = new ShopifyProductService({
      manager: MockManager,
      shopifyClientService: ShopifyClientServiceMock,
      productService: ProductServiceMock,
      shopifyRedisService: ShopifyRedisServiceMock,
      shippingProfileService: ShippingProfileServiceMock,
      productVariantService: ProductVariantServiceMock,
    })

    beforeEach(async () => {
      jest.clearAllMocks()
    })

    it("succesfully creates a product from Shopify", async () => {
      const data = shopifyProducts.new_ipod

      const product = await shopifyProductService.create(data)

      expect(ShopifyRedisServiceMock.shouldIgnore).toHaveBeenCalledTimes(1)
      expect(ShopifyRedisServiceMock.addIgnore).toHaveBeenCalledTimes(1)
      expect(ShippingProfileServiceMock.retrieveDefault).toHaveBeenCalledTimes(
        1
      )
      expect(ProductServiceMock.create).toHaveBeenCalledTimes(1)
      expect(ProductVariantServiceMock.create).toHaveBeenCalledTimes(4)
      expect(product).toEqual(
        expect.objectContaining({
          id: "prod_ipod",
        })
      )
    })
  })

  describe("update", () => {
    const shopifyProductService = new ShopifyProductService({
      manager: MockManager,
      shopifyClientService: ShopifyClientServiceMock,
      productService: ProductServiceMock,
      shopifyRedisService: ShopifyRedisServiceMock,
      shippingProfileService: ShippingProfileServiceMock,
      productVariantService: ProductVariantServiceMock,
    })

    beforeEach(async () => {
      jest.clearAllMocks()
    })

    it("updates a product and adds 4 new variants", async () => {
      const data = shopifyProducts.ipod_update
      jest
        .spyOn(shopifyProductService, "addProductOptions_")
        .mockImplementation(() => ({
          ...medusaProducts.ipod,
          options: [
            ...medusaProducts.ipod.options,
            {
              id: "opt_01FHZ9ZFCPCQ7B1MPDD9X9YQX4",
              title: "Memory",
              product_id: "prod_01FHZ9ZFC3KKYKA35NXNJ5A7FR",
              created_at: "2021-10-14T11:46:10.391Z",
              updated_at: "2021-10-14T11:46:10.391Z",
              deleted_at: null,
              metadata: null,
            },
          ],
        }))

      await shopifyProductService.update(medusaProducts.ipod, data)

      expect(ProductVariantServiceMock.update).toHaveBeenCalledTimes(8)
      expect(ProductServiceMock.update).toHaveBeenCalledTimes(1)
    })

    it("updates a product and deletes 2 existing variants", async () => {
      const data = { ...shopifyProducts.ipod, id: "shopify_deleted" }
      data.variants = data.variants.slice(1, -1)
      await shopifyProductService.update(
        { ...medusaProducts.ipod, id: "shopify_deleted" },
        data
      )

      expect(ProductVariantServiceMock.delete).toHaveBeenCalledTimes(2)
      expect(ProductServiceMock.update).toHaveBeenCalledTimes(1)
    })
  })
})
