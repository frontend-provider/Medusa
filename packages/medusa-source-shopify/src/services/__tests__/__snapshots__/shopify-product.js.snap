// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopifyProductService normalizeProduct_ succesfully normalizes a product from Shopify 1`] = `
Object {
  "description": "<p>It's the small iPod with one very big idea: Video. Now the world's most popular music player, available in 4GB and 8GB models, lets you enjoy TV shows, movies, video podcasts, and more. The larger, brighter display means amazing picture quality. In six eye-catching colors, iPod nano is stunning all around. And with models starting at just $149, little speaks volumes.</p>",
  "external_id": "shopify_ipod",
  "handle": "ipod-nano",
  "images": Array [
    "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano.png?v=1633120966",
    "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano-2.png?v=1633120966",
    "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano.png?v=1633120966",
  ],
  "is_giftcard": false,
  "metadata": Object {
    "vendor": "Apple",
  },
  "options": Array [
    Object {
      "title": "Color",
      "values": Array [
        Object {
          "value": "Pink",
        },
        Object {
          "value": "Red",
        },
        Object {
          "value": "Green",
        },
        Object {
          "value": "Black",
        },
      ],
    },
  ],
  "product_type": Object {
    "value": "Cult Products",
  },
  "status": "proposed",
  "tags": Array [
    Object {
      "value": "Emotive",
    },
    Object {
      "value": " Flash Memory",
    },
    Object {
      "value": " MP3",
    },
    Object {
      "value": " Music",
    },
  ],
  "thumbnail": "https://cdn.shopify.com/s/files/1/0006/9093/3842/products/ipod-nano.png?v=1633120966",
  "title": "IPod Nano - 8GB",
  "variants": Array [
    Object {
      "allow_backorder": true,
      "barcode": "1234_pink",
      "inventory_quantity": 10,
      "manage_inventory": true,
      "metadata": Object {
        "sh_id": 808950810,
      },
      "options": Array [
        Object {
          "value": "Pink",
        },
      ],
      "prices": Array [
        Object {
          "amount": 19900,
          "currency_code": "usd",
        },
      ],
      "sku": "IPOD2008PINK",
      "title": "Pink",
      "upc": "1234_pink",
      "variant_rank": 1,
      "weight": 567,
    },
    Object {
      "allow_backorder": true,
      "barcode": "1234_red",
      "inventory_quantity": 20,
      "manage_inventory": true,
      "metadata": Object {
        "sh_id": 49148385,
      },
      "options": Array [
        Object {
          "value": "Red",
        },
      ],
      "prices": Array [
        Object {
          "amount": 19900,
          "currency_code": "usd",
        },
      ],
      "sku": "IPOD2008RED",
      "title": "Red",
      "upc": "1234_red",
      "variant_rank": 2,
      "weight": 567,
    },
    Object {
      "allow_backorder": true,
      "barcode": "1234_green",
      "inventory_quantity": 30,
      "manage_inventory": true,
      "metadata": Object {
        "sh_id": 39072856,
      },
      "options": Array [
        Object {
          "value": "Green",
        },
      ],
      "prices": Array [
        Object {
          "amount": 19900,
          "currency_code": "usd",
        },
      ],
      "sku": "IPOD2008GREEN",
      "title": "Green",
      "upc": "1234_green",
      "variant_rank": 3,
      "weight": 567,
    },
    Object {
      "allow_backorder": true,
      "barcode": "1234_black",
      "inventory_quantity": 40,
      "manage_inventory": true,
      "metadata": Object {
        "sh_id": 457924702,
      },
      "options": Array [
        Object {
          "value": "Black",
        },
      ],
      "prices": Array [
        Object {
          "amount": 19900,
          "currency_code": "usd",
        },
      ],
      "sku": "IPOD2008BLACK",
      "title": "Black",
      "upc": "1234_black",
      "variant_rank": 4,
      "weight": 567,
    },
  ],
}
`;
